mountpoint-s3-client/src/imds_crt_client.rs

```diff
@@ -0,0 +1,109 @@
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+
+use futures::channel::oneshot;
+use futures::Future;
+use mountpoint_s3_crt::auth::imds_client::ImdsClient;
+use mountpoint_s3_crt::auth::imds_client::ImdsClientConfig;
+use mountpoint_s3_crt::common::allocator::Allocator;
+use mountpoint_s3_crt::common::error;
+use mountpoint_s3_crt::io::channel_bootstrap::{ClientBootstrap, ClientBootstrapOptions};
+use mountpoint_s3_crt::io::event_loop::EventLoopGroup;
+use mountpoint_s3_crt::io::host_resolver::{HostResolver, HostResolverDefaultOptions};
+use mountpoint_s3_crt::io::retry_strategy::{RetryStrategy, StandardRetryOptions};
+use pin_project::pin_project;
+use thiserror::Error;
+
+#[derive(Debug)]
+/// Instance Metadata Service (IMDS) client responsible for sending ec2 instance metadata query requests.
+pub struct ImdsCrtClient {
+    imds_client: ImdsClient,
+    event_loop_group: EventLoopGroup,
+    allocator: Allocator,
+}
+
+impl ImdsCrtClient {
+    pub fn new() -> Result<Self, error::Error> {
+        let allocator = Allocator::default();
+
+        let mut event_loop_group = EventLoopGroup::new_default(&allocator, None, || {}).unwrap();
+
+        let resolver_options = HostResolverDefaultOptions {
+            max_entries: 8,
+            event_loop_group: &mut event_loop_group,
+        };
+
+        let mut host_resolver = HostResolver::new_default(&allocator, &resolver_options).unwrap();
+
+        let bootstrap_options = ClientBootstrapOptions {
+            event_loop_group: &mut event_loop_group,
+            host_resolver: &mut host_resolver,
+        };
+
+        let client_bootstrap = ClientBootstrap::new(&allocator, &bootstrap_options).unwrap();
+
+        let mut retry_strategy_options = StandardRetryOptions::default(&mut event_loop_group);
+        let retry_strategy = RetryStrategy::standard(&allocator, &retry_strategy_options).unwrap();
```

Comment:
1. Doesn't need to be `mut`.
2. What are the default options? Are they appropriate for the IMDS use case?

Reply:
1.fixed

Reply:
2. fixed